### PR TITLE
4764 debian evr version comparator

### PR DESCRIPTION
**Pull request by DProvinciani**
|Related issue|
|---|
|[4769](https://github.com/wazuh/wazuh/issues/4769)|

## Description

This change implements a mechanism to have a version comparator for each type of supported packages. It currently implements a comparison mechanism for `deb` and `rpm` packages versions. Both implementations are a direct code porting from the `DPKG` and `rpmLib` projects respectively. For the case of packages that make use of the **NVD**, it makes use of the `deb` version comparator as it was in the past.

### CODE REVIEW

File: src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c [hunk 1]
```diff
@@ -152,9 +183,186 @@ const char *__wrap_sqlite3_errmsg(){
...
+    version_a->epoch = 1;
+    version_a->version = "";
+    version_a->revision = "";
+    version_b->epoch = 0;
+    version_b->version = "";
+    version_b->revision = "";
+
+    int ret = pkg_version_relate(version_a, PKG_RELATION_NONE, version_b, VER_TYPE_NONE);
+    assert_int_equal(ret, 1);
+    ret = pkg_version_relate(version_a, PKG_RELATION_GT, version_b, VER_TYPE_DEB);
```
Comment (TomasTurina) [inline src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c:202]: For cases like this one, you would create different tests to make sure you are testing only one particular situation.

Comment (DProvinciani) [reply src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c:202]: So, do you mean that I should create a different test case for each `PKG_RELATION_XXX` where `version_a->epoch = 1` and `version_b->epoch = 0` ??

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+    rc = comparator(a->version, b->version);
+    if (rc)
+        return rc;
+
+    return comparator(a->revision, b->revision);
+}
+
+/**
+ * Debian EVR string comparator. Extracted from Debian source code.
```
Comment (chemamartinez) [inline src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c:417]: Please change this description to adapt it to the new purpose of the function (to choose the comparator and call it)

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    /* compared identically but the segment sepparating characters were */
+    /* different */
+    if ((!*one) && (!*two)) return 0;
+
+    /* whichever version still has characters left over wins */
+    if (!*one) return -1; else return 1;
+}
+
+/**
+ * Compares two Debian versions. Extracted from Debian source code.
```
Comment (chemamartinez) [inline src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c:387]: It not used for Debian only after the latest changes, update the description.

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+        return c;
+    else if (c == '~')
+        return -1;
+    else if (c)
+        return c + 256;
+    else
+        return 0;
+}
+
+static int deb_verrevcmp(const char *a, const char *b)
```
Comment (chemamartinez) [inline src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c:214]: Doxygen comments are missing for this function. Can you add them?

File: src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c (hunk 1 above)
Comment (DProvinciani) [reply src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c]: DONE

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+    switch (vertype)
+    {
+    case VER_TYPE_DEB:
+        comparator = deb_verrevcmp; break;
+    case VER_TYPE_RPM:
+        comparator = rpm_vercmp; break;
+    case VER_TYPE_NVD:
+        comparator = deb_verrevcmp; break; // TODO: Temporary assigning the debian comparator until having the one for NVD
+    default:
+        merror("unknown version_type %d", vertype);
```
[the remark this answers is not in the document]
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c]: DONE

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+
+    // Assigning version comparator
+    switch (vertype)
+    {
+    case VER_TYPE_DEB:
+        comparator = deb_verrevcmp; break;
+    case VER_TYPE_RPM:
+        comparator = rpm_vercmp; break;
+    case VER_TYPE_NVD:
+        comparator = deb_verrevcmp; break; // TODO: Temporary assigning the debian comparator until having the one for NVD
```
[the remark this answers is not in the document]
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c:452]: DONE

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c (hunk 2 above)
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c]: DONE

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c (hunk 3 above)
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c]: DONE

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c (hunk 4 above)
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector_evr.c:221]: DONE

File: src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c [hunk 7]
```diff
@@ -152,9 +183,294 @@ const char *__wrap_sqlite3_errmsg(){
...
+
+    version_a->epoch = 0;
+    version_a->version = "5.3.0.45";
+    version_a->revision = "";
+    version_b->epoch = 0;
+    version_b->version = "4.18.1";
+    version_b->revision = "";
+
+    int ret = pkg_version_relate(version_a, PKG_RELATION_GT, version_b, VER_TYPE_DEB);
+    assert_int_equal(ret, 1);
```
Comment (TomasTurina) [inline src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c:335]: Could you please do the same with this test and the three tests below?

Comment (DProvinciani) [reply src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c:335]: DONE

**Pull request merged**